tape/src/main/java/com/squareup/tape/QueueFile.java

```diff
@@ -462,6 +459,31 @@ public synchronized void forEach(ElementVisitor reader) throws IOException {
     }
   }
 
+  /**
+   * First reads header to update element counts and file length. Then Scans each element in the
+   * queue for its length. To be used for Integrity Scans to detect queue corruption.
+   *
+   * @return int[] of all segment lengths.
+   * @throws IOException if there is a problem scanning the file (assume corruption).
+   */
+  public synchronized int[] scanAll() throws IOException {
+    // Read header to update all our counts
+    readHeader();
+    int position = first.position;
+    final int[] lengths = new int[elementCount];
+    int totalSize = HEADER_LENGTH;
+    for (int i = 0; i < elementCount; i++) {
+      Element current = readElement(position);
+      lengths[i] = current.length;
+      totalSize += current.length + Element.HEADER_LENGTH;
+      position = wrapPosition(current.position + Element.HEADER_LENGTH + current.length);
```

Comment:
wrapPosition performs no check, and assume valid length. Here, if we read a length of 1.7Gb, that'll wrap around but position will still be larger than the file size. readElement will cause another wrapPosition (still not enough) and then we'll attempt reading out of the file from the RAF so that should throw.

Reply:
Yup, fixed that now.